src/c/types/array.c

```diff
@@ -69,332 +66,96 @@ void ucdr_buffer_to_array(ucdrBuffer* ub, uint8_t* array, const uint32_t size, c
+        } \
+    } \
     return !ub->error;
-}
 
-bool ucdr_deserialize_array_byte_1(ucdrBuffer* ub, uint8_t* array, const uint32_t size)
-{
-    ucdr_buffer_to_array(ub, array, size, 1);
-    return !ub->error;
-}
+#define UCDR_SERIALIZE_ARRAY_BYTE_2(TYPE, ENDIAN) UCDR_SERIALIZE_ARRAY_BYTE_N(TYPE, 2, ENDIAN)
+#define UCDR_SERIALIZE_ARRAY_BYTE_4(TYPE, ENDIAN) UCDR_SERIALIZE_ARRAY_BYTE_N(TYPE, 4, ENDIAN)
+#define UCDR_SERIALIZE_ARRAY_BYTE_8(TYPE, ENDIAN) UCDR_SERIALIZE_ARRAY_BYTE_N(TYPE, 8, ENDIAN)
 
-bool ucdr_deserialize_array_byte_2(ucdrBuffer* ub, const ucdrEndianness endianness, uint16_t* array, const uint32_t size)
-{
-    ub->iterator += ucdr_buffer_alignment(ub, 2);
-
-    if(UCDR_MACHINE_ENDIANNESS == endianness)
-    {
-        ucdr_buffer_to_array(ub, (uint8_t*)array, size * 2, 2);
-    }
-    else
-    {
-        for(uint32_t i = 0; i < size; i++)
-        {
-            ucdr_deserialize_byte_2(ub, endianness, array + i);
-        }
-    }
+#define UCDR_SERIALIZE_ARRAY_DEFINITION(SUFFIX, TYPE, SIZE) \
+    bool ucdr_serialize_array ## SUFFIX(ucdrBuffer* ub, const TYPE * array, const uint32_t size) \
+    { \
+        UCDR_SERIALIZE_ARRAY_BYTE_ ## SIZE(TYPE, ub->endianness) \
+    } \
+    bool ucdr_serialize_endian_array ## SUFFIX(ucdrBuffer* ub, const ucdrEndianness endianness, const TYPE * array, const uint32_t size) \
+    { \
+        UCDR_SERIALIZE_ARRAY_BYTE_ ## SIZE(TYPE, endianness) \
+    } \
 
+// -------------------------------------------------------------------
+//                         DESERIALIZE MACROS
+// -------------------------------------------------------------------
+#define UCDR_DESERIALIZE_ARRAY_BYTE_1(TYPE, ENDIAN) \
+    (void)ENDIAN; \
+    ucdr_buffer_to_array(ub, (uint8_t*)array, size, 1); \
     return !ub->error;
-}
-
-bool ucdr_deserialize_array_byte_4(ucdrBuffer* ub, const ucdrEndianness endianness, uint32_t* array, const uint32_t size)
-{
-    ub->iterator += ucdr_buffer_alignment(ub, 4);
-
-    if(UCDR_MACHINE_ENDIANNESS == endianness)
-    {
-        ucdr_buffer_to_array(ub, (uint8_t*)array, size * 4, 4);
-    }
-    else
-    {
-        for(uint32_t i = 0; i < size; i++)
-        {
-            ucdr_deserialize_byte_4(ub, endianness, array + i);
-        }
-    }
 
+#define UCDR_DESERIALIZE_ARRAY_BYTE_N(TYPE, SIZE, ENDIAN) \
+    ub->iterator += ucdr_buffer_alignment(ub, (uint32_t)SIZE); \
+    if(UCDR_MACHINE_ENDIANNESS == ENDIAN) \
+    { \
+        ucdr_buffer_to_array(ub, (uint8_t*)array, size * SIZE, SIZE); \
+    } \
+    else \
+    { \
+        for(uint32_t i = 0; i < size; ++i) \
+        { \
+            ucdr_deserialize_endian_ ## TYPE(ub, ENDIAN, array + i); \
+        } \
+    } \
     return !ub->error;
-}
 
-bool ucdr_deserialize_array_byte_8(ucdrBuffer* ub, const ucdrEndianness endianness, uint64_t* array, const uint32_t size)
-{
-    ub->iterator += ucdr_buffer_alignment(ub, 8);
+#define UCDR_DESERIALIZE_ARRAY_BYTE_2(TYPE, ENDIAN) UCDR_DESERIALIZE_ARRAY_BYTE_N(TYPE, 2, ENDIAN)
```

Comment:
There are an associated headers to this functions in `_internal.h` that could be write as macros too, I think

Reply:
In fact, we don't need the internal headers now. I removed it.

Reply:
You removed `array_internals.h` and `sequence_internals.h`, does the `basic_internals.h` be removed too? I do not find any explicit call to its functions (but maybe is because of the macros)

Reply:
Removed also `basic_internal.h`.